src/LessMsi.Cli/Program.cs

```diff
@@ -95,25 +101,81 @@ public static int Main(string[] args)
 			}
 		}
 
-		/// <summary>
-		/// Extracts all files contained in the specified .msi file into the specified output directory.
-		/// </summary>
-		/// <param name="msiFileName">The path of the specified MSI file.</param>
-		/// <param name="outDirName">The directory to extract to. If empty it will use the current directory.</param>
-		/// <param name="filesToExtract">The files to be extracted from the msi. If empty all files will be extracted.</param>
-		public static void DoExtraction(string msiFileName, string outDirName, List<string> filesToExtract )
-		{
-			if (string.IsNullOrEmpty(outDirName))
-				outDirName = Path.GetFileNameWithoutExtension(msiFileName);
-			EnsureFileRooted(ref msiFileName);
-			EnsureFileRooted(ref outDirName);
+        /// <summary>
+        /// Extracts all files contained in the specified .msi file into the specified output directory.
+        /// </summary>
+        /// <param name="msiFileName">The path of the specified MSI file.</param>
+        /// <param name="outDirName">The directory to extract to. If empty it will use the current directory.</param>
+        /// <param name="filesToExtract">The files to be extracted from the msi. If empty all files will be extracted.</param>
+        /// /// <param name="extractionMode">Enum value for files extraction without folder structure</param>
+        public static void DoExtraction(string msiFileName, string outDirName, List<string> filesToExtract, eExtractionMode extractionMode)
+        {
+            if (string.IsNullOrEmpty(outDirName))
+                outDirName = Path.GetFileNameWithoutExtension(msiFileName);
+
+            EnsureFileRooted(ref msiFileName);
+            EnsureFileRooted(ref outDirName, extractionMode);
 
             var msiFile = new LessIO.Path(msiFileName);
 
-			Console.WriteLine("Extracting \'" + msiFile + "\' to \'" + outDirName + "\'.");
+            Console.WriteLine("Extracting \'" + msiFile + "\' to \'" + outDirName + "\'.");
 
-			Wixtracts.ExtractFiles(msiFile, outDirName, filesToExtract.ToArray(), PrintProgress);
-		}
+            Wixtracts.ExtractFiles(msiFile, outDirName, filesToExtract.ToArray(), PrintProgress);
+
+            if (isExtractionModeFlat(extractionMode))
+            {
+                Dictionary<string, int> fileNameCountingDict = new Dictionary<string, int>();
+
+                string originalOutDirName = outDirName.Remove(outDirName.Length - c_TempFolderSuffix.Length) + "\\";

```

Comment:
It seems unnecessarily complicated to compute this new suffix name in one method (in `EnsureFileRooted`) and then to make assumption about that logic in this method later. So instead, I recommend that you just compute the outdir name in this `DoExtraction` method above and remove that logic from `EnsureFileRooted`. Then before you overrwrite the `outDir` local variable here, you can preserve it's original value in `originalOutDirName` in the same method. I think that will simplify the logic here. Let me know if I'm missing something!

Reply:
I agree with your approach here and will change the code accordingly.